arch/x86/pagetables.c

```diff
@@ -159,42 +163,65 @@ void *vmap(void *va, mfn_t mfn, unsigned int order, unsigned long flags) {
     spin_lock(&lock);
 
 #if defined(__x86_64__)
-    l3t_mfn = get_pgentry_mfn(get_cr3_mfn(&cr3), l4_table_index(va), L4_PROT_USER);
+    l3t_mfn = get_pgentry_mfn(get_cr3_mfn(&cr3), l4_table_index(va), l4_flags);
 #else
     l3t_mfn = get_cr3_mfn(&cr3);
 #endif
 
     if (order == PAGE_ORDER_1G) {
         tab = init_map_mfn(l3t_mfn);
         entry = &tab[l3_table_index(va)];
-        set_pgentry(entry, mfn, flags | _PAGE_PSE);
+        set_pgentry(entry, mfn, l3_flags | _PAGE_PSE);
         goto done;
     }
 
-    l2t_mfn = get_pgentry_mfn(l3t_mfn, l3_table_index(va), L3_PROT_USER);
+    l2t_mfn = get_pgentry_mfn(l3t_mfn, l3_table_index(va), l3_flags);
 
     if (order == PAGE_ORDER_2M) {
         tab = init_map_mfn(l2t_mfn);
         entry = &tab[l2_table_index(va)];
-        set_pgentry(entry, mfn, flags | _PAGE_PSE);
+        set_pgentry(entry, mfn, l2_flags | _PAGE_PSE);
         goto done;
     }
 
-    l1t_mfn = get_pgentry_mfn(l2t_mfn, l2_table_index(va), L2_PROT_USER);
+    l1t_mfn = get_pgentry_mfn(l2t_mfn, l2_table_index(va), l2_flags);
 
     tab = init_map_mfn(l1t_mfn);
     entry = &tab[l1_table_index(va)];
-    set_pgentry(entry, mfn, flags);
+    set_pgentry(entry, mfn, l1_flags);
 
 done:
     spin_unlock(&lock);
     return va;
 }
 
+void *vmap(void *va, mfn_t mfn, unsigned int order, unsigned long flags) {
+    switch (order) {
+    case PAGE_ORDER_1G:
+        return vmap_ext(va, mfn, order, L4_PROT_USER, flags, 0, 0);
+    case PAGE_ORDER_2M:
+        return vmap_ext(va, mfn, order, L4_PROT_USER, L3_PROT_USER, flags, 0);
+    default:
+        return vmap_ext(va, mfn, order, L4_PROT_USER, L3_PROT_USER, L2_PROT_USER, flags);
+    }
+}
+
```

Comment:
Quite honestly I have a mixed feelings about this change. The `vmap` function is (supposed to be) an internal coarse-grained interface. Maybe we do not need the split into `vmap_ext()` and `vmap()`, but we could make the `vmap()` using the four flag parameters directly.

Or, we could create separate helper function for each of the supported orders. But, I do not like this approach, cause it will limit flexibility of the interface.